### DISCUSSION TITLE
Backend design with Prisma & Apollo

### DISCUSSION THREAD

**Discussion by lensbart** [prisma/prisma#12643] [Q&A]
Hello,

I’m about to begin a migration from Sequelize to Prisma. The [migration guide](https://www.prisma.io/docs/guides/migrate-to-prisma/migrate-from-sequelize) is a terrific resource and allowed me to set up Prisma (side-by-side with Sequelize) in no time. Thank you!

Now, it’s time to gradually convert all queries. Before I start doing so, I wanted to take the chance to reorganise some of our backend logic so as to be more robust and in line with best practices. Here, I have a couple of questions:

1. Currently, all of our resolver logic is written in the resolver files themselves. I’m told that it’s a better practice to keep the API layer thin, and move the logic for authorization and retrieval to a “domain” or “data access” layer. Moving logic into functions in a separate folder is easy enough, but I’d like to avoid having to pass `context` (the third argument in the Apollo resolver function signature, which contains information about the logged-in user) every time. E.g.:

```ts
// Logic in resolver:
export const books = async (_, __, { prisma, user }) => {
  // ... authorization checks ...
  return prisma.book.findMany({ where: { userId: user.id } })
}
```

vs.

```ts
// in a folder called “domain”
class User {
  // could be reused by other resolvers & has access to private methods
  books() {
    // ... authorization checks ...
    return this.prisma.book.findMany({ where: { userId: this.context.userId } })
  }
}

// “thin” API layer:
export const books = async () => User.books()
```

So my first question would be: is this overkill? It seems to be standard practice in enterprise apps to separate business logic from the API resolvers. But nearly all GraphQL tutorials put the business logic directly inside the resolver (for the sake of simplicity?).

Secondly: what would be the best way to avoid having to pass context information (currently logged in user, `prisma`) to the “domain” methods? [This comment](https://github.com/prisma/prisma/issues/5017#issuecomment-605004652) describes an approach based on [Apollo Data Sources](https://www.apollographql.com/docs/apollo-server/data/data-sources). I’m inclined to follow this approach, with separate classes (instead of one `Store`) to split up my domain.

Third: is the combination of Prisma with Apollo Data Sources redundant? I [this blog post by an Apollo team member](https://www.apollographql.com/blog/backend/data-sources/a-deep-dive-on-apollo-data-sources/#coding-our-own-server), they list Apollo Data Sources and Prisma as alternatives.

If you think there are better, alternative approaches to splitting domain logic from the API layer, feel free to suggest otherwise!

Thank you!

**Comment by pantharshit00** [comment]
Hello @lensbart, our apologies for the delay in getting back to you! Due to the large number of questions in our community channels, unfortunately this slipped through the cracks. It's very important for us to address your question, so let us know in a comment if this is still an open discussion. 

> So my first question would be: is this overkill? It seems to be standard practice in enterprise apps to separate business logic from the API resolvers. But nearly all GraphQL tutorials put the business logic directly inside the resolver (for the sake of simplicity?).

It is upto your preference. Many frameworks like Nest actually prefer this approach.


> what would be the best way to avoid having to pass context information (currently logged in user, prisma) to the “domain” methods? https://github.com/prisma/prisma/issues/5017#issuecomment-605004652 describes an approach based on [Apollo Data Sources](https://www.apollographql.com/docs/apollo-server/data/data-sources). I’m inclined to follow this approach, with separate classes (instead of one Store) to split up my domain.

I think you should make them contextless and pass the data by hand in resolvers. That way you might be able to share this domain code for a second server application if need occurs. 


> Third: is the combination of Prisma with Apollo Data Sources redundant? I [this blog post by an Apollo team member](https://www.apollographql.com/blog/backend/data-sources/a-deep-dive-on-apollo-data-sources/#coding-our-own-server), they list Apollo Data Sources and Prisma as alternatives.

I also think apollo data source will be redundant. It will be just a useless wrapper in this case.

**Reply by vladi-stevanovic** [reply]
:wave: Hello @lensbart , I’m going through pending GitHub Discussions to check if you need any help. Is this issue still open?

Please let us know if we can investigate it further or if you've already solved it, we'd love to know how!

It would be super useful for other users facing this problem! :raised_hands:

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
